llvm/include/llvm/Target/GlobalISel/Combine.td

```diff
@@ -1005,7 +1005,13 @@ def sdiv_by_const : GICombineRule<
    [{ return Helper.matchSDivByConst(*${root}); }]),
   (apply [{ Helper.applySDivByConst(*${root}); }])>;
 
-def intdiv_combines : GICombineGroup<[udiv_by_const, sdiv_by_const]>;
+def sdiv_by_pow2 : GICombineRule<
+  (defs root:$root),
+  (match (wip_match_opcode G_SDIV):$root,
+   [{ return Helper.matchSDivByPow2(*${root}); }]),
+  (apply [{ Helper.applySDivByPow2(*${root}); }])>;
+
+def intdiv_combines : GICombineGroup<[udiv_by_const, sdiv_by_const, sdiv_by_pow2]>;
```

Comment:
`sdiv_by_pow2` matches a true subset of `sdiv_by_const`. Is there some guarantee which rules matches?
The SelectionDAG pattern matcher uses lexcial order, which can be influenced with a priority. I did not see anything about it in the [docs](https://llvm.org/docs/GlobalISel/MIRPatterns.html), but I may have overlooked it.

Reply:
Yeah, that's a good point. I maybe need to put the logic into `sdiv_by_const`.

Reply:
I would prefer to have it separate, since my backend has no support for multiplying 2 numbers with the result being twice as large. Then `sdiv_by_const` results in code which is more expensive than using the `sdiv` instruction. However, `sdiv_by_pow2` is very useful (I considered to implement it myself, so thank you for working on that).
 
The matching order is a separate problem, I wonder if there are already other dependencies between rules.